grpc/dns_resolver.go

```diff
@@ -0,0 +1,53 @@
+package grpc
+
+import (
+	"context"
+	"net"
+
+	"google.golang.org/grpc/naming"
+)
+
+// dnsResolver implements both the naming.Resolver and naming.Watcher
+// interfaces. It's a temporary shim until we upgrade to the latest gRPC, which
+// has a built-in DNS resolver. It looks up the hostname only once; it doesn't
+// monitor for changes.
+type dnsResolver struct {
+	host, port string
+	// ch is used to enforce the "lookup only once" behavior.
+	ch chan bool
+}
+
+func newDNSResolver(host, port string) *dnsResolver {
+	return &dnsResolver{
+		host: host,
+		port: port,
+		ch:   make(chan bool, 1),
+	}
+}
+
+func (dr *dnsResolver) Resolve(target string) (naming.Watcher, error) {
+	return dr, nil
+}
+
+// Next is called in a loop by grpc.RoundRobin expecting updates. We provide a
+// single update then block forever.
+func (dr *dnsResolver) Next() ([]*naming.Update, error) {
+	// Stick a value on the channel, which has capacity 1. This will succed once,
+	// then each subsequent call will block forever.
+	dr.ch <- true
+	addrs, err := net.DefaultResolver.LookupHost(context.TODO(), dr.host)
```

Comment:
Why `context.TODO`? Should probably just be a `context.Background()` or a `context.WithTimeout`, or a comment should be added explaining what the use of `context.TODO` means.